paddle/fluid/framework/details/scope_buffered_monitor.cc

```diff
@@ -0,0 +1,202 @@
+  }
+}
+
+static size_t GetTensorMemorySize(Scope *scope, bool clear_cpu_tensor) {
+  std::unordered_set<Tensor *> tensor_set;
+  GetTensors(scope, &tensor_set);
+  size_t memory_size = 0;
+  std::unordered_set<memory::Allocation *> allocation_set;
+  for (auto *tensor : tensor_set) {
+    if (clear_cpu_tensor && platform::is_cpu_place(tensor->place())) {
+      tensor->clear();
+    } else {
+      auto allocation = tensor->Holder().get();
+      if (!allocation_set.count(allocation)) {
+        memory_size += allocation->size();
+        allocation_set.insert(allocation);
+      }
+    }
+  }
+  return memory_size;
+}
+
+size_t GetScopeVarMemorySize(Scope *scope) {
+  return GetTensorMemorySize(scope, false /*clear_cpu_tensor*/);
+}
+
+ScopeBufferedMonitor::ScopeBufferedMonitor(
+    const std::vector<platform::Place> &places,
+    const std::vector<Scope *> &local_exec_scopes)
+    : places_(places), local_exec_scopes_(local_exec_scopes) {
+  pre_local_exec_scopes_.resize(local_exec_scopes_.size());
+  post_local_exec_scopes_.resize(local_exec_scopes_.size());
+}
+
+void ScopeBufferedMonitor::Apply(const std::function<void()> &callback,
+                                 bool has_fetch) {
+  std::unique_ptr<platform::RecordEvent> pre_local_exec_scopes_event(
+      new platform::RecordEvent(
+          "ScopeBufferedMonitor::pre_local_exec_scopes_process"));
+  for (size_t scope_id = 0; scope_id < local_exec_scopes_.size(); ++scope_id) {
+    pre_local_exec_scopes_.at(scope_id).clear();
+    auto scopes = local_exec_scopes_.at(scope_id)->kids();
+    VLOG(10) << "pre_local_exec_scopes[" << scope_id
+             << "] sub-scope: " << scopes.size();
+    pre_local_exec_scopes_.at(scope_id).insert(scopes.begin(), scopes.end());
+  }
+  pre_local_exec_scopes_event.reset();
+
+  callback();
+
+  std::unique_ptr<platform::RecordEvent> post_local_exec_scopes_event(
+      new platform::RecordEvent(
+          "ScopeBufferedMonitor::post_local_exec_scopes_process"));
+  for (size_t scope_id = 0; scope_id < local_exec_scopes_.size(); ++scope_id) {
+    post_local_exec_scopes_.at(scope_id).clear();
+    auto scopes = local_exec_scopes_.at(scope_id)->kids();
+    VLOG(10) << "post_local_exec_scopes[" << scope_id
+             << "] sub-scope: " << scopes.size();
+    post_local_exec_scopes_.at(scope_id).insert(scopes.begin(), scopes.end());
+  }
+
+  history_local_exec_scopes_.emplace_back();
+  auto &incr_local_exec_scopes = history_local_exec_scopes_.back();
+  incr_local_exec_scopes.resize(local_exec_scopes_.size());
+  for (size_t scope_id = 0; scope_id < local_exec_scopes_.size(); ++scope_id) {
+    for (auto &scope : post_local_exec_scopes_.at(scope_id)) {
+      if (!pre_local_exec_scopes_.at(scope_id).count(scope)) {
+        incr_local_exec_scopes.at(scope_id).insert(scope);
+      }
+    }
+
+    if (incr_local_exec_scopes.at(scope_id).size() &&
+        FLAGS_local_exe_sub_scope_limit > 0) {
+      LOG_FIRST_N(WARNING, 1)
```

Comment:
Do we need to add this warning to users? Users may not know what it means.

Reply:
Ok, I will fix this in next PR.

Reply:
Done.